src/test/java/edu/ucla/library/libcal/handlers/ProxyHandlerTest.java

```diff
@@ -0,0 +1,92 @@
+
+package edu.ucla.library.libcal.handlers;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import info.freelibrary.util.HTTP;
+import info.freelibrary.util.Logger;
+import info.freelibrary.util.LoggerFactory;
+
+import edu.ucla.library.libcal.Config;
+import edu.ucla.library.libcal.Constants;
+import edu.ucla.library.libcal.MessageCodes;
+import edu.ucla.library.libcal.verticles.MainVerticle;
+
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.junit.jupiter.api.TestInstance.Lifecycle;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import io.vertx.core.DeploymentOptions;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.Vertx;
+import io.vertx.ext.web.client.HttpResponse;
+import io.vertx.ext.web.client.WebClient;
+import io.vertx.ext.web.client.predicate.ResponsePredicate;
+import io.vertx.ext.web.codec.BodyCodec;
+import io.vertx.junit5.VertxExtension;
+import io.vertx.junit5.VertxTestContext;
+
+/**
+ * Tests {@link ProxyHandler#handle}.
+ */
+@ExtendWith(VertxExtension.class)
+@TestInstance(Lifecycle.PER_CLASS)
+public class ProxyHandlerTest {
+
+    /**
+     * The test's logger.
+     */
+    private static final Logger LOGGER = LoggerFactory.getLogger(ProxyHandlerTest.class, MessageCodes.BUNDLE);
+
+    /**
+     * The default port that the application listens on.
+     */
+    private static String DEFAULT_PORT = "8888";
+
+    /**
+     * Sets up the test.
+     *
+     * @param aVertx A Vert.x instance
+     * @param aContext A test context
+     */
+    @BeforeAll
+    public void setUp(final Vertx aVertx, final VertxTestContext aContext) {
+        final int port = Integer.parseInt(System.getenv().getOrDefault(Config.HTTP_PORT, DEFAULT_PORT));
+        final DeploymentOptions options = new DeploymentOptions();
+
+        options.setConfig(new JsonObject().put(Config.HTTP_PORT, port));
+
+        aVertx.deployVerticle(MainVerticle.class.getName(), options).onSuccess(result -> aContext.completeNow())
+                .onFailure(aContext::failNow);
+    }
+
+    /**
+     * Tests that a client can get LibCap API output.
+     *
+     * @param aVertx A Vert.x instance
+     * @param aContext A test context
+     */
+    @Test
+    public void testGetOutput(final Vertx aVertx, final VertxTestContext aContext) {
+        final String requestPath = "/libcal/api%2F1.1%2Fhours%2F2572";
```

Comment:
I don't think the library website should have to percent-encode request URIs, but rather should be able to call this service almost exactly as if it were talking to LibCal directly (except that a bearer token isn't required). I.e., I would expect this test to pass with the following change applied:

```suggestion
        final String requestPath = "/api/1.1/hours/2572";
```

Although, I'm curious why you took the percent-encoding route.

Reply:
Because `api%2F1.1%2Fhours%2F2572` is interpreted as one variable; `api/1.1/hours/2572` is interpreted as four segments of the URL path, which doesn't match the OpenAPI spec for ProxyHandler. (Do a `vertx:run` and `curl` each version to see behavior.) The choices seem to be to 1) URL encode the API query; 2) switch to using POST, where the query variable wouldn't need to be encoded; 3) make a separate route handler for each LibCal API. URL encoding was fastest to program, and I'm guessing not that hard for Apps team to handle (though we should ask them).

Reply:
It's possible to just forward all incoming requests (other than `/status`), unmodified, to ProxyHandler. You can get a "catch-all" route with [`Router::route()`](https://vertx.io/docs/apidocs/io/vertx/ext/web/Router.html#route--) (we're currently creating a router instance with `routeBuilder.createRouter()` in MainVerticle) and then [register ProxyHandler with it](https://vertx.io/docs/apidocs/io/vertx/ext/web/Route.html#handler-io.vertx.core.Handler-). In that case, ProxyHandler wouldn't be associated with any OpenAPI operations, but since we canceled the ticket for creating an OpenAPI document, I think that's fine, especially if it means Apps can just use the same API as LibCal.

Reply:
Tried doing an empty `route()`, and a [route by path beginning with something](https://vertx.io/docs/vertx-web/java/#_routing_by_paths_that_begin_with_something)... but now I'm getting either a "resource not found" or "not implemented" error when I call the handler via `curl`. Perhaps I'm not implementing the route in the right place in the verticle?

Reply:
I think part of the problem is that you're creating two router instances: one [here](https://github.com/UCLALibrary/libcal-proxy/pull/7/commits/cfcfacd981e897613fae691507f5ed220759a42d#diff-2d4f9a42f6ef9e77bdb08b7508a366b402a9af3a5b0bf3ce13ec2564d4bf2de5R94) (which is associated with ProxyHandler), then a new one [here](https://github.com/UCLALibrary/libcal-proxy/pull/7/commits/cfcfacd981e897613fae691507f5ed220759a42d#diff-2d4f9a42f6ef9e77bdb08b7508a366b402a9af3a5b0bf3ce13ec2564d4bf2de5R96) (which is not).